algorithms/src/std_algorithms/Kokkos_NonModifyingSequenceOperations.hpp

```diff
@@ -0,0 +1,2100 @@
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_NON_MODIFYING_SEQUENCE_OPERATIONS_HPP
+#define KOKKOS_NON_MODIFYING_SEQUENCE_OPERATIONS_HPP
+
+#include <Kokkos_Core.hpp>
+#include "Kokkos_BeginEnd.hpp"
+#include "Kokkos_Constraints.hpp"
+#include "Kokkos_MinMaxOperations.hpp"
+#include "Kokkos_ModifyingOperations.hpp"
+#include "Kokkos_HelperPredicates.hpp"
+#include "Kokkos_RandomAccessIterator.hpp"
+#include <string>
+
+namespace Kokkos {
+namespace Experimental {
+namespace Impl {
+
+// ------------------------------------------
+//
+// functors
+//
+// ------------------------------------------
+
+template <bool is_find_if, class IndexType, class IteratorType,
+          class ReducerType, class PredicateType>
+struct StdFindIfOrNotFunctor {
+  using red_value_type = typename ReducerType::value_type;
+
+  IteratorType m_first;
+  ReducerType m_reducer;
+  PredicateType m_p;
+
+  KOKKOS_FUNCTION
+  void operator()(const IndexType i, red_value_type& red_value) const {
+    const auto& my_value = m_first[i];
+
+    // if doing find_if, look for when predicate is true
+    // if doing find_if_not, look for when predicate is false
+    const bool found_condition = is_find_if ? m_p(my_value) : !m_p(my_value);
+
+    auto rv =
+        found_condition
+            ? red_value_type{i}
+            : red_value_type{::Kokkos::reduction_identity<IndexType>::min()};
+
+    m_reducer.join(red_value, rv);
+  }
+
+  KOKKOS_FUNCTION
+  StdFindIfOrNotFunctor(IteratorType first, ReducerType reducer,
+                        PredicateType p)
+      : m_first(first), m_reducer(std::move(reducer)), m_p(std::move(p)) {}
```

Comment:
Global comment:  It doesn't hurt to also do `m_first(std::move(first))`, since we are doing that for all the other parameters.

I've noticed this isn't consistent.  For instance, the `StdEqualFunctor` constructor doesn't move any of its parameters.

Reply:
Our idea was to move all predicates (as potentially "heavier" objects), I must have missed `StdEqualFunctor`.